### DISCUSSION TITLE
Am I doing something wrong? Have I discovered a bug? wifi.scan() doesn't seem to work as expected.

### DISCUSSION THREAD

**Discussion by wh33t** [micropython/micropython#10886] [ESP32]
So my apologies upfront if this whole discussion is bogus due to my noobness. 

I am completely unfamiliar with ESP32 and Python ... I'm just trying to get my feet wet with the fundamentals.

I am using the Thing Plus C ESP32 from Sparkfun and the Thonny Editor.

I just want to power up the ESP32, and scan to see which wifi networks it can see.

This code will not produce any output when run.

```
# wifi_test.py
import network
import time

# Create a station object to store our connection
wifi = network.WLAN(network.STA_IF)
wifi.active(True)  

available_wifi_networks = wifi.scan()

for i in available_wifi_networks:
    print (i[0] + " " + i[1])
```

However, I have noticed that if I declare the wifi object, set it's active status to false, sleep the device for 1 second, and then set it back to active it will work as expected and reliably so.

```
# wifi_test.py
import network
import time

# Create a station object to store our connection
wifi = network.WLAN(network.STA_IF)
wifi.active(False)  
time.sleep(1)
wifi.active(True)  

available_wifi_networks = wifi.scan()

for i in available_wifi_networks:
    print (i[0] + " " + i[1])
```
Is this ESP32 unit from Sparkfun faulty? Am I doing something wrong here?

Any and all help greatly appreciated. Cheers.

**Comment by davefes** [comment]
In your first block of code and after `wifi.active(True)` try this:
```py
 #  seems more reliable to start with a fresh connect()
    if sta_if.isconnected():
        sta_if.disconnect()
        print('started in the connected state, but now disconnected')
    else:
        print('started in the disconnected state')

    utime.sleep(0.1)
```

**Comment by wh33t** [comment]
```
# wifi_test.py
import network
import time
import utime

# Create a station object to store our connection
wifi = network.WLAN(network.STA_IF)
wifi.active(True)

#  seems more reliable to start with a fresh connect()
if wifi.isconnected():
    wifi.disconnect()
    print('started in the connected state, but now disconnected')
else:
    print('started in the disconnected state')

utime.sleep(0.1)

available_wifi_networks = wifi.scan()

for i in available_wifi_networks:
    print ("> " + str(i[0],"utf-8"))
```

Outputs

> started in the disconnected state

**Reply by davefes** [reply]
Have been using f strings for the last month on the ESP32.

Hmm your whole email shows up in my Gmail message but not here??
Error line 13 often means that the error is in the line before.

**Reply by wh33t** [reply]
I'm totally new to python and esp32 so I'm not sure what the issue might be. Just created and ran this script.

```
# fstring_test.py
list = ['value1','value2']

print(f"{list}")
```
Output
> >>> %Run -c $EDITOR_CONTENT
> Traceback (most recent call last):
>   File "<stdin>", line 4
> SyntaxError: invalid syntax
> >>>

**Comment by davefes** [comment]
Maybe :
```py
try:
     available_wifi_networks = wifi.scan()
    print(f'{available_wifi_networks}')
except OSError as error:
    print(f'{error}')

**Comment by wh33t** [comment]
```
# wifi_test.py
import network
import time
import utime

# Create a station object to store our connection
wifi = network.WLAN(network.STA_IF)
wifi.active(True)

try:
    available_wifi_networks = wifi.scan()
    for i in available_wifi_networks:
        print(f'{available_wifi_networks}')
except OSError as error:
    print(f"{error}")
```

> Traceback (most recent call last):
>   File "<stdin>", line 13
> SyntaxError: invalid syntax

[I don't think f strings are supported in micropython](https://docs.micropython.org/en/latest/genrst/core_language.html)

If I remove the f strings

```
# wifi_test.py
import network
import time
import utime

# Create a station object to store our connection
wifi = network.WLAN(network.STA_IF)
wifi.active(True)

try:
    available_wifi_networks = wifi.scan()
    for i in available_wifi_networks:
        print(available_wifi_networks)
except OSError as error:
    print(error)
```
Produces no output

**Reply by davefes** [reply]
Maybe:
```py

available_wifi_networks() = wifi.scan()
```

**Reply by wh33t** [reply]
> Maybe:
> 
> ```python
> available_wifi_networks() = wifi.scan()
> ```

```
# wifi_test.py
import network
import time
import utime

# Create a station object to store our connection
wifi = network.WLAN(network.STA_IF)
wifi.active(True)

available_wifi_networks() = wifi.scan()

for i in available_wifi_networks:
    print ("> " + str(i[0],"utf-8"))
```
Output

> Traceback (most recent call last):
>   File "<stdin>", line 10, in <module>
> SyntaxError: can't assign to expression

Line 10
`available_wifi_networks() = wifi.scan()`

I'm starting to wonder if I maybe flashed the wrong version of micropython. From the firmwares available here: https://micropython.org/download/esp32/ which one should I be using? Do I want one with IDF?

**Reply by davefes** [reply]
Which ESP32 have you got?  The images are built using different version of ESP-IDF use the v4 unless you have a reason to go back in history.

**Reply by davefes** [reply]
I looked at declaring an empty tuple.  Then I would expect you then write into it.

```py
available_wifi_networks()
available_wifi_networks = wifi.scan()
```
https://www.kdnuggets.com/2019/11/python-tuples-methods.html

**Reply by wh33t** [reply]
> Which ESP32 have you got? The images are built using different version of ESP-IDF use the v4 unless you have a reason to go back in history.

[Sparkfun Thing Plus C](https://www.sparkfun.com/products/20168) I have the latest IDF (v1.14 (2021-02-02) .bin) flashed on the device right now.

**Reply by davefes** [reply]
My script above doesn't work either.

On a new ESP32 using Micropython I first erase the flash then download the image here:
https://micropython.org/download/esp32/
and the first .bin file
[v1.19.1 (2022-06-18) .bin](https://micropython.org/resources/firmware/esp32-20220618-v1.19.1.bin)
Am I on the same page?

**Reply by davefes** [reply]
> Scanning is only possible on STA interface. Returns list of tuples with the information about WiFi access points:

    (ssid, bssid, channel, RSSI, security, hidden)

Didn't see a list of tuples :(

The magic to get output from scan escapes me, sorry.

**Reply by wh33t** [reply]
> My script above doesn't work either.
> 
> On a new ESP32 using Micropython I first erase the flash then download the image here: https://micropython.org/download/esp32/ and the first .bin file [v1.19.1 (2022-06-18) .bin](https://micropython.org/resources/firmware/esp32-20220618-v1.19.1.bin) Am I on the same page?

I updated my previous comment. I am actually using v1.14 (2021-02-02) .bin which is an IDF 3.x firmware. Maybe that is the problem here. I wasn't sure what IDF 3.x meant, and I think I get it now. The downloads listed under the heading "Firmware" are the IDF version 4 Is that correct? If so I am one IDF version behind. I certainly have no reason that I'm aware of to be on outdated firmware.

**Reply by wh33t** [reply]
> > Scanning is only possible on STA interface. Returns list of tuples with the information about WiFi access points:
> 
> ```
> (ssid, bssid, channel, RSSI, security, hidden)
> ```
> 
> Didn't see a list of tuples :(
> 
> The magic to get output from scan escapes me, sorry.

All good. I can make it work by setting it's .active state to false, sleeping for 1 second, and then setting it's active state back to true. I just wasn't sure why it wasn't working as I expected. I will flash to the most up to date firmware and see if the issue persists. You've been a great help. Thank you so much.

**Reply by davefes** [reply]
That could be one problem I am using v1.19.1, but don't know how to handle the output!

**Reply by wh33t** [reply]
> That could be one problem I am using v1.19.1, but don't know how to handle the output!

I think just 

`print(tuple_or_list)` will dump it all out console.

**Reply by davefes** [reply]
I have been working with WiFi on the ESP32 for over a year and I am still discovering work-arounds to get consistent WiFi connections.  Some of my early attempts are littered with sleeps.  Try reducing that sleep and see when it falls over. Maybe it is a bug, sorry I headed down the wrong track.

**Reply by davefes** [reply]
Falls over at less than 10ms.

**Reply by wh33t** [reply]
> Falls over at less than 10ms.

Interesting. So as long as there is 10ms of sleep it seems to reliably bring up the wifi?

**Reply by davefes** [reply]
It does for the one WiFi in my house.  All bets are off for anyone else's WiFi.

**Reply by wh33t** [reply]
> It does for the one WiFi in my house. All bets are off for anyone else's WiFi.

I am now discovering I don't even need the sleep any longer, I just need to set to inactive, and then back to active.

```
# wifi_test.py
import network
import time

wifi = network.WLAN(network.STA_IF)
wifi.active(False)
wifi.active(True)

available_wifi_networks = wifi.scan()

for i in available_wifi_networks:
    print ("> " + str(i[0],"utf-8"))
```

Seems to work every time.

**Reply by davefes** [reply]
Same here.  You found the key!  There must be a reasonable explanation for this.  With 'sleep_ms(.1)` it fell-over.

**Reply by wh33t** [reply]
> Same here. You found the key! There must be a reasonable explanation for this. With 'sleep_ms(.1)` it fell-over.

I gotta say. The world of embedded development seems like a huge crapshoot so far lol. Even just to locate accurate and up to date tutorials and guides and specs is really frustrating. I guess that's the price you pay for wanting to create electronic gadgets though. 

Alright, so I updated to the latest firmware, the issue still persists on my end. So I'm guessing it's just some weird ghost in the machine and I presume we should file a bug report somewhere? I'm not crazy for thinking that you shouldn't have to set to the wifi object to inactive, then to active just to scan for wifi networks right? That has to be a bug?

**Reply by davefes** [reply]
I do a lot of work with ESPNow and Glenn20 has some utils.  Looking at wifi.py I see in the reset method something similar.
https://github.com/glenn20/micropython-espnow-utils/tree/main/src
line 119

**Reply by davefes** [reply]
Setting up a static IP also requires some hoops to jump through.  I find the ESP32 a really useful device but there things happening down at a lower level that I wouldn't have a chance in sorting out.

I think there is a way to raise an issue from the thread itself.  I haven't been bold-enough  to do it.

**Reply by davefes** [reply]
@wh33t,

Could you try Glenn20's wifi.py and if that works properly I'll have a go at raising an issue?

**Reply by davefes** [reply]
On 2nd thought that is not going to work as scan on the WiFi network is not implemented.

**Reply by wh33t** [reply]
I am moving on from this issue, I will use the work around we've discovered. 

I do think we should get this thread in front of anyone who works on the esp32 micropython port. It's obviously not a huge issue, but maybe we should let them decide whether or not it's a bug that needs fixing. Do you know how to get someone like that to see the thread?

**Reply by davefes** [reply]
Raised an issue

**Comment by Yanqi2023** [comment]
You should pressed En button，on the board。That can make wifi on。
